src/target/spirv/codegen_spirv.cc

```diff
@@ -561,6 +630,55 @@ void CodeGenSPIRV::VisitStmt_(const ForNode* op) {
 
   // Loop head
   builder_->StartLabel(head_label);
+
+  // In a normal matmul, the update semantics c += a * b is implemented by load and store to the
+  // memory location corresponding to the scalar c. But when C, A, and B are cooperative matrices,
+  // C += A * B cannot be implemented in the same way, since they cannot be load from or stored in
+  // a buffer.
+
+  // We leverage SPIRV's SSA semantics to implement multiply-add update on C matrices.
+  // The original buffer is subdivided into fixed-size matrices, and multiply-add on individual
+  // matrix is unrolled. For example, if the buffer for the C matrix is of shape 64x64, we
+  // materialze 16 16x16 matrices and generate 16 multiply-add instructions. Each matrix is
+  // represented by a phi value, and each iteration of the reduction loop reads the phi values for
+  // the C matrices and update them.
+
+  // A map from a buffer with kCooperativeMatrixNV storage scope (whose size can be arbitrary)
+  // to cooperative matrices "C" (whose sizes are fixed, e.g. 16x16).
+  // Each matrix is identified by an element offset into the buffer. This encoding lets us bridge
+  // the gap between a typical TIR matmul schedule (which uses
+  // `cache_read(..., "cooperative_matrix_nv")` from arbitrary-sized shared memory) and
+  // the fixed-size matrices as expected by the extension.
+  std::unordered_map<const VarNode*, std::unordered_set<int>> accum_matrices;
+
+  if (op->kind == ForKind::kSerial) {
+    // If this is a serial loop, record which C matrices are updated in this loop.
+    tir::PostOrderVisit(op->body, [&accum_matrices](const ObjectRef& obj) {
```

Comment:
Or, are you imagining a scenario where the matrix to be accumulated is initialized from `cooperative_matrix_load`, rather than zero initialized inside the kernel before the reduction loop?

Reply:
The issue I'm imagining would occur when the reduction loop occurs within another loop.  For example, suppose I took the matrix multiplication unit test, but added a batch dimension (rough sketch below).  The cooperative matrix loads wouldn't depend on the batch number for their shapes, and the declaration/allocation of the cooperative matrices could occur within the body of `batch_i`.  However, the loop over `batch_i` shouldn't have an `OpPhi` associated with the cooperative matrices, because there is no value of the matrix prior to entering the loop over `batch_i`.

```python
@T.prim_func
def main(
    X: T.Buffer((4, 16, 32), "float16"),
    W: T.Buffer((4, 32, 16), "float16"),
    compute: T.Buffer((4, 16, 16), "float32"),
):
    for batch_i in range(4):
        X_shared = T.decl_buffer((16, 32), "float16", scope="shared")
        W_shared = T.decl_buffer((32, 16), "float16", scope="shared")
        X_shared_cooperative_matrix_nv = T.decl_buffer(
            (16, 32), "float16", scope="cooperative_matrix_nv"
        )
        W_shared_cooperative_matrix_nv = T.decl_buffer(
            (32, 16), "float16", scope="cooperative_matrix_nv"
        )
        compute_cooperative_matrix_nv = T.decl_buffer(
            (16, 16), scope="cooperative_matrix_nv"
        )
        i_0_j_0_fused = T.launch_thread("blockIdx.x", 1)
        with T.launch_thread("threadIdx.x", 32) as tx:
            T.cooperative_matrix_fill_NV(
                compute_cooperative_matrix_nv.data, 0, 16, 16, T.float32(0)
            )
        # Continues with remainder of matrix multiplication
        for k_0 in range(2):
            ...
```